UnityProject/Assets/Scripts/Items/Weapons/SawnOff.cs

```diff
@@ -0,0 +1,100 @@
+using System;
+using System.Collections.Generic;
+using Items;
+using AddressableReferences;
+using Mirror;
+using UnityEditor;
+using UnityEngine;
+using UnityEngine.Serialization;
+using Weapons;
+
+[RequireComponent(typeof(ItemAttributesV2))]
+[RequireComponent(typeof(Gun))]
+public class SawnOff : MonoBehaviour, ICheckedInteractable<InventoryApply>
+{
+
+	private ItemAttributesV2 itemAttComp;
+	private Gun gunComp;
+	private SpriteHandler spriteHandler;
+
+	private bool isSawn = false;
+
+	[SerializeField, Tooltip("Determines if the gun will explode in the users face when sawn while it still contains ammo")]
+	private bool ammoBackfire = true;
+
+	[SerializeField, Tooltip("Sawn off item size")]
+	private ItemSize sawnSize = ItemSize.Medium;
+	
+	[SerializeField, Tooltip("Value that determines how far a shot will deviate. (Never set this higher then 0.5 unless you want questionable results.)")]
+	private float sawnMaxRecoilVariance;
+
+	[SerializeField, Tooltip("Recoil camera config to switch to when sawn off")]
+	private CameraRecoilConfig SawnCameraRecoilConfig;
+
+	private void Awake()
+	{
+		spriteHandler = GetComponentInChildren<SpriteHandler>();
+		itemAttComp = gameObject.GetComponent<ItemAttributesV2>();
+		gunComp = gameObject.GetComponent<Gun>();
+		spriteHandler.ChangeSprite(0);
+	}
+
+	public bool WillInteract(InventoryApply interaction, NetworkSide side)
+	{
+		if (DefaultWillInteract.Default(interaction, side) == false) return false;
+		if (side == NetworkSide.Server && DefaultWillInteract.Default(interaction, side)) return true;
+
+		//only reload if the gun is the target and tool is in hand slot
+		if (interaction.TargetObject == gameObject && interaction.IsFromHandSlot && side == NetworkSide.Client)
+		{
+			//TODO: switch this trait to the circular saw when that is implemented
+			if (Validations.HasItemTrait(interaction.UsedObject, CommonTraits.Instance.Welder))
+			{
+				return true;
+			}
+		}
+		return false;
+	}
+
+	public void ServerPerformInteraction(InventoryApply interaction)
+	{
+		if (interaction.TargetObject == gameObject && interaction.IsFromHandSlot)
+		{
+			//TODO: switch this trait to the circular saw when that is implemented
+			if (Validations.HasItemTrait(interaction.UsedObject, CommonTraits.Instance.Welder) && gunComp.FireCountDown == 0)
+			{
+				if (isSawn)
+				{
+					Chat.AddExamineMsg(interaction.Performer, $"The {gameObject.ExpensiveName()} is already shortened!");
+				}
+
+				WaitFor.Seconds(0.25f);
```

Comment:
does this work? You shouldn't be able to stop execution outside an async/coroutine